### PR TITLE
refactor issue & cleanup gadgets

### PR DESCRIPTION
**Pull request by DavidBadura**
gadgets & issue refactoring

### CODE REVIEW

File: src/Gadget/PhpmdGadget.php [hunk 1]
```diff
@@ -98,7 +96,8 @@ private function convertFromXmlToArray($xml)
      */
     private function createIssue($file, array $data)
     {
-        $issue = new Issue(trim($data['#']), self::NAME, Issue::LEVEL_WARNING);
+        $issue = new Issue(trim($data['#']));
```
Comment (tolry) [inline src/Gadget/PhpmdGadget.php]: I do not really like the way, we are currently instantiating issues right now. On most gadgets we have the same combination of

``` php
$issue = new Issue('one');
$issue->Level();
$issue->setFile();
$issue->setLine();
```
I would suggest using named constructors with appropriate names. http://verraes.net/2014/06/named-constructors-in-php/ Any opinions on this?

e.g.

``` php
$issue = Issue::createWithinGadgetIncludingFileAndLine(...);
```

**Comment by tolry** [comment]
This saves a repetitive task within the gadgets, but it is a bit magic. There is this issue entity that has a property `gadgetName` that can be set by the gadget. But whatever is set by the gadget, will be ignored / overwritten later on. This might be confusing to new developers.

You would save this redundancy also by simply adding a `addIssue()` method to the AbstractGadget, or am I missing something? Regarding my previous post above: we could also give shortcuts in AbstractGadget for  additional parameters, e.g. `addIssueIncludingFilename()`

**Comment by DavidBadura** [comment]
`setGadget` is marked as a internal method with the comment `will be set by executor`. Modern IDE's display this information and marked this method.

Why should the gadget set the name? Here can happen errors. For example false name so we can not map the issue to the gadget anymore. I don't see any benefits if the gadget can set the name self. Only problems and redundancy.

File: src/Gadget/PhpmdGadget.php (hunk 1 above)
Comment (DavidBadura) [inline src/Gadget/PhpmdGadget.php]: i find this better:

```php
$issue = Issue::createWithCodeSnippet($message, $file, $line, $level);
```

**Comment by DavidBadura** [comment]
and i don't like add a lot of methods in AbstractGadget for shortcuts if the method don't need class attributes for her logic. So the method can be anywhere. For example `findFiles` what i moved into a seperate util class. `findFiles` have nothing to do with the gadget and it was protected, so we can't unit test it.

**Comment by tolry** [comment]
But "addIssue" is the perfect example for a method that we can put in the AbstractGadget class. It's directly related to gadgets an helps adding issues from a gadget. I really think it's a bad idea to have this property, which will be 'brutally" overwritten by another class later. Marking the method as "internal" is also not correct, since it is not used internally but by the executor.

I agree that adding the gadget string manually is error prone, the most explicit version would be to pass the gadget instance to the issue and the issue class decides how this relation is resolved (by calling getName() on the gadget object).

Let's talk about this tomorrow, in person? Maybe the rest has some opinion on this too. Let's fight on the whiteboard :wink:

File: src/Gadget/PhpmdGadget.php (hunk 1 above)
Comment (tolry) [inline src/Gadget/PhpmdGadget.php]: :+1: I like the wording

**Comment by DavidBadura** [comment]
:+1: for pass the gadget instance to the issue. the arguments have convinced :wink:

**Comment by tolry** [comment]
Great! Will you make the changes in this branch or will you (or somebody else) make a fresh branch?

**Comment by DavidBadura** [comment]
i will refresh this branch

**Comment by DavidBadura** [comment]
ready

**Comment by tolry** [comment]
I still think, named constructors are a good idea to remove the repetitive 

``` php
        $issue = new Issue($this->gadget, $message);
        $issue->setLevel($level);
        $issue->setFile($file);
```

but we can refactor that later.

:+1: 

Anybody else care to give a :+1: so we can merge? :smile:

**Comment by DanielBadura** [comment]
:+1:

**Comment by tolry** [comment]
so bullied your brother into :+1:'ing? :wink:

**Pull request merged**